### ISSUE TITLE
Dice coefficient code does not sort

### ISSUE THREAD

**Issue by nbgl** [n1analytics/anonlink#135]
The current Dice coefficient code sorts pairs by (i0, -score, i1), where i0 and i1 are the indices of the first and second record in the pair respectively.

For example (5, .8, 10) will come before (6, .9, 10), even though the second pair has a higher similarity. The greedy solver prefers pairs that come first, so in this case it would prefer the first matching. Logically though, we should prefer the second matching since (all else being equal) the second pair is a better match for record 10 in the second dataset. This outcome can be achieved by sorting all pairs in order of similarity instead.

Improper sorting can make a big difference depending on the noise in the dataset. This effect is shown in the following plot:
![greedy-sorting](https://user-images.githubusercontent.com/12507513/43578651-c5af3998-9692-11e8-9ded-eac18d435718.png)

With noise of 0.1 (excuse the arbitrary scale), we go from balanced accuracy of 0.880 to 0.945. And with noise of 0.032, we go from 0.960 to 0.987. That's a big difference.

I've implemented this change in Anonlink in a fairly naive way. (PR coming soon, mainly so I don't forget about it. I won't get offended if it doesn't get merged.) Benchmarks show that this change is barely noticeable for less noisy problems (ones with higher threshold), but significantly slows down similarity pair generation for noisier problems (lower threshold). That said, the noisier tasks are also the ones that benefit the most from this change, so it may be worth the cost.

### Before
```
Threshold: 0.5
| Sizes   | Total Time | Throughput  |
|         | (s)        | (1e6 cmp/s) |
|---------|------------|-------------|
| 1e3×1e3 |       0.11 |        9.03 |
| 2e3×2e3 |       0.45 |        8.77 |
| 3e3×3e3 |       1.01 |        8.89 |
| 4e3×4e3 |       1.82 |        8.76 |
| 5e3×5e3 |       2.85 |        8.77 |
| 6e3×6e3 |       4.13 |        8.70 |
| 7e3×7e3 |       5.63 |        8.69 |
| 8e3×8e3 |       7.47 |        8.56 |
| 9e3×9e3 |       9.47 |        8.55 |
| 1e4×1e4 |      11.77 |        8.49 |
| 2e4×2e4 |      49.29 |        8.11 |
| 3e4×3e4 |     112.88 |        7.97 |
| 4e4×4e4 |     208.64 |        7.66 |

Threshold: 0.7
| Sizes   | Total Time | Throughput  |
|         | (s)        | (1e6 cmp/s) |
|---------|------------|-------------|
| 1e3×1e3 |       0.01 |       80.01 |
| 2e3×2e3 |       0.04 |       92.33 |
| 3e3×3e3 |       0.09 |       96.85 |
| 4e3×4e3 |       0.15 |      102.59 |
| 5e3×5e3 |       0.24 |      102.97 |
| 6e3×6e3 |       0.33 |      106.85 |
| 7e3×7e3 |       0.44 |      109.01 |
| 8e3×8e3 |       0.61 |      103.27 |
| 9e3×9e3 |       0.81 |       99.33 |
| 1e4×1e4 |       1.00 |       99.65 |
| 2e4×2e4 |       3.65 |      109.43 |
| 3e4×3e4 |       8.21 |      109.52 |
| 4e4×4e4 |      16.25 |       98.45 |
| 5e4×5e4 |      24.60 |      101.61 |
| 6e4×6e4 |      38.30 |       93.97 |
| 7e4×7e4 |      58.23 |       84.14 |
| 8e4×8e4 |      69.12 |       92.58 |
| 9e4×9e4 |      85.44 |       94.80 |
```

### After
```
Threshold: 0.5
| Sizes   | Total Time | Throughput  |
|         | (s)        | (1e6 cmp/s) |
|---------|------------|-------------|
| 1e3×1e3 |       0.14 |        6.74 |
| 2e3×2e3 |       0.57 |        6.99 |
| 3e3×3e3 |       1.31 |        6.82 |
| 4e3×4e3 |       2.33 |        6.86 |
| 5e3×5e3 |       3.83 |        6.51 |
| 6e3×6e3 |       5.54 |        6.48 |
| 7e3×7e3 |       7.81 |        6.26 |
| 8e3×8e3 |      10.00 |        6.39 |
| 9e3×9e3 |      12.85 |        6.30 |
| 1e4×1e4 |      16.02 |        6.23 |
| 2e4×2e4 |      71.64 |        5.58 |
| 3e4×3e4 |     197.61 |        4.55 |
| 4e4×4e4 |     451.96 |        3.54 |

Threshold: 0.7
| Sizes   | Total Time | Throughput  |
|         | (s)        | (1e6 cmp/s) |
|---------|------------|-------------|
| 1e3×1e3 |       0.04 |       22.99 |
| 2e3×2e3 |       0.06 |       66.24 |
| 3e3×3e3 |       0.10 |       83.72 |
| 4e3×4e3 |       0.17 |       90.87 |
| 5e3×5e3 |       0.28 |       88.55 |
| 6e3×6e3 |       0.36 |       97.57 |
| 7e3×7e3 |       0.50 |       96.43 |
| 8e3×8e3 |       0.62 |      102.15 |
| 9e3×9e3 |       0.83 |       96.52 |
| 1e4×1e4 |       0.98 |      101.79 |
| 2e4×2e4 |       3.80 |      105.17 |
| 3e4×3e4 |       8.26 |      108.89 |
| 4e4×4e4 |      14.99 |      106.73 |
| 5e4×5e4 |      24.46 |      102.17 |
| 6e4×6e4 |      36.41 |       98.87 |
| 7e4×7e4 |      52.50 |       93.32 |
| 8e4×8e4 |      68.89 |       92.89 |
| 9e4×9e4 |      87.17 |       92.91 |
```
## What now
Three options:
1. Ignore this problem completely
2. Put effort into coming up with a faster way to sort
3. Merge the naive solution and improve performance if we experience problems

![solver-diagram](https://user-images.githubusercontent.com/12507513/43582054-61c6b79c-969e-11e8-85a6-1b229f03ac34.png)

@unzvfu, any thoughts?

**Comment by unzvfu** [comment]
I pick 3.

**Comment by nbgl** [comment]
@sjhardy pointed out that the framework I’ve been using to test solvers does not behave _quite_ like a real similarity matrix. The main difference seems to be that my fake similarities don’t necessarily respect the triangle inequality (which gives us a lower bound on _sim_(_a_, _c_) if we know the _sim_(_a_, _b_) and _sim_(_b_, _c_).

The accuracy improvements are indeed visible on similarities computed from CLKs. I used data generated by the [GeCo](https://dmm.anu.edu.au/geco/) tool set to fairly high levels of noise. The magnitude of improvement depended on the threshold chosen. For one choice of threshold, sorting the similarities improved the balanced accuracy from .91 (precision .51, recall .83) to .95 (precision .55, recall .90). For another choice of threshold the balanced accuracy changed from .90 (precision .37, recall .80) to .98 (precision .45, recall .96). Plots below:

![accuracy](https://user-images.githubusercontent.com/12507513/43686619-454dc89c-990c-11e8-87f4-8c3904cbb4e9.png)
![precision-v-recall](https://user-images.githubusercontent.com/12507513/43686618-4514ea0e-990c-11e8-9628-e69da53d66fe.png)

Thresholds below .5 are excluded from the above plots because that’d take too long to compute.

**Comment by hardbyte** [comment]
Ditto option 3 seems to make the most sense for now. 

> Make it right, then (optional for experts) make it fast.

**Comment by nbgl** [comment]
Fixed in #136.
